pkg/operator/csi/csistorageclasscontroller/csi_storageclass_controller.go

```diff
@@ -0,0 +1,148 @@
+// overwriting value that user might have manually changed.
+// If the asset file does not have defaultScAnnotationKey set at all, controller
+// just skips any checks and modifications and applies the StorageClass as it is.
+// It produces following Conditions:
+// StorageClassControllerDegraded - failed to apply StorageClass provided
+type CSIStorageClassController struct {
+	name               string
+	assetFunc          resourceapply.AssetFunc
+	file               string
+	kubeClient         kubernetes.Interface
+	storageClassLister v1.StorageClassLister
+	operatorClient     v1helpers.OperatorClient
+	eventRecorder      events.Recorder
+}
+
+func NewCSIStorageClassController(
+	name string,
+	assetFunc resourceapply.AssetFunc,
+	file string,
+	kubeClient kubernetes.Interface,
+	informerFactory informers.SharedInformerFactory,
+	operatorClient v1helpers.OperatorClient,
+	eventRecorder events.Recorder) factory.Controller {
+	c := &CSIStorageClassController{
+		name:               name,
+		assetFunc:          assetFunc,
+		file:               file,
+		kubeClient:         kubeClient,
+		storageClassLister: informerFactory.Storage().V1().StorageClasses().Lister(),
+		operatorClient:     operatorClient,
+		eventRecorder:      eventRecorder,
+	}
+
+	return factory.New().WithSync(
+		c.Sync,
+	).ResyncEvery(
+		time.Minute,
+	).WithSyncDegradedOnError(
+		operatorClient,
+	).WithInformers(
+		operatorClient.Informer(),
+		informerFactory.Storage().V1().StorageClasses().Informer(),
+	).ToController(
+		"StorageClassController",
+		eventRecorder,
+	)
+}
+
+func (c *CSIStorageClassController) Sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	klog.V(4).Infof("StorageClassController sync started")
+	defer klog.V(4).Infof("StorageClassController sync finished")
+
+	opSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+	if opSpec.ManagementState != operatorapi.Managed {
+		return nil
+	}
+
+	syncErr := c.syncStorageClass(ctx)
+
+	return syncErr
+}
+
+func (c *CSIStorageClassController) syncStorageClass(ctx context.Context) error {
+	expectedScBytes, err := c.assetFunc(c.file)
+	if err != nil {
+		return err
+	}
+
+	expectedSC := resourceread.ReadStorageClassV1OrDie(expectedScBytes)
+
+	existingSCs, err := c.storageClassLister.List(labels.Everything())
+	if err != nil {
+		klog.V(2).Infof("could not list StorageClass objects")
+		return err
+	}
+
+	defaultSCCount := 0
+	var existingSCAnnotation string
+
+	// Skip the default SC annotation check if it's not in the input/expectedSC.
+	if expectedSC.Annotations != nil && expectedSC.Annotations[defaultScAnnotationKey] != "" {
+		for _, sc := range existingSCs {
+			if sc.Annotations[defaultScAnnotationKey] == "true" && sc.Name != expectedSC.Name {
+				defaultSCCount++
+			}
+			if sc.Name == expectedSC.Name && sc.Annotations != nil {
+				existingSCAnnotation = sc.Annotations[defaultScAnnotationKey]
```

Comment:
If the annotation exists and has empty value (`""`), `existingSCAnnotation` will be `""` and `expectedSC.Annotations[defaultScAnnotationKey]` won't be cleared at `if existingSCAnnotation` below.

Reply:
Right...let's add even more unit tests then. Posted a fix now.